redaxo/src/core/lib/sql/sql.php

```diff
@@ -1392,6 +1392,29 @@ public function escapeLikeWildcards(string $value): string
         return str_replace(['_', '%'], ['\_', '\%'], $value);
     }
 
+    /**
+     * Escapes and transforms values for `IN (...)` clause.
+     *
+     * Example: `$sql->setQuery('SELECT * FROM my_table WHERE foo IN ('.$sql->in($values).')');`
+     *
+     * @param scalar[] $values
```

Comment:
vllt präziser:

```suggestion
     * @param list<string|int|bool> $values
```

Reply:
floats wollte ich eigentlich nicht verbieten. da sie hier aber recht unwahrscheinlich sind, habe ich sie nicht gesondert behandelt, sondern werden zu strings gecastet.

Ob die Daten als list reinkommen, oder array mit anderen keys (zum beispiel mit lücken in der nummerierung), ist hier eigentlich egal.
Meine Erfahrung ist, das `list<...>` als Parameter nerven kann, da man an der aufrufenden Stelle öfters für psalm nicht ersichtlich eine saubere Liste hat.

Reply:
ok wg. `list`. float in IN hab ich bisher nie gesehen.. sich dass mysql das kann?

Reply:
Ich denke schon, dass floats funktionieren. Praktischen Nutzen sehe ich dafür aber auch nur sehr begrenzt.
Wobei ich bools in `IN()` eigentlich noch unsinniger finde.
Letztlich ist es mir eigentlich egal, ob wir sie akzeptieren oder nicht.

https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#operator_in

Wobei ich hier gerade noch was interessantes gelesen habe, dass man in der Liste nicht ints und strings mischen soll.
Jetzt weiß ich nicht, ob wir daher einfach grundsätzlich in strings umwandeln sollen. Oder ob es sich lohnt, erst zu prüfen, ob es ein reines int-array ist, und dann nicht in strings zu wandeln.
int-arrays sind bei `IN()` ja recht üblich, und wenn die column integer ist, möchte man ja `int_col IN ('1', '2')` eigentlich auch nicht, weil mysql dann ja die typen auch wieder angleichen muss.

Reply:
Laut phpdoc wird nun entweder ein int-array oder ein string-array erwartet.
bei anderen typen kommt eine exception. 
bei int/string-mix wandelt der code die ints in strings um, auch wenn laut phpdoc gar kein mix vorgesehen ist.

Reply:
da ich heute zufällig (in einer anderen software) drüber gestolpert bin.

mysql IN mit string werten ist wichtig für enum spalten, da diese - auch wenn numerische werte drinnen stehen - immer mit quotes abgefragt werden müssen (sonst wird z.b. `-1` mit `1` vermischt, wenn `IN(1)` statt `IN("1")` genutzt wird

Reply:
Wenn man enums mit integers vergleicht, dann wird glaube ich der Index der Enums herangezogen.
Das kann theoretisch gewollt sein, gerade bei nicht nummerischen Enums.

Jedenfalls denke ich, dass der Caller unserer Methode in deinem Fall die Werte als String übergeben muss.
Ich denke nicht, dass wir deswegen grundsätzlich Integers in Strings wandeln sollten.

Reply:
MySQL rät aber auch sehr davon ab, numerische Enums zu verwenden:

https://dev.mysql.com/doc/refman/8.0/en/enum.html#enum-limits

> We strongly recommend that you do not use numbers as enumeration values, because...

Reply:
ich wollte nicht sagen dass wir integers in strings umwandeln sollten.
nur auf den fall hinweisen weil ich zufällig heute drüber gestolpert bin.